src/query/catalog/src/plan/projection.rs

```diff
@@ -49,6 +52,28 @@ impl Projection {
             Projection::InnerColumns(path_indices) => schema.inner_project(path_indices),
         }
     }
+
+    pub fn project_column_leaves<'a>(
+        &'a self,
+        column_leaves: &'a ColumnLeaves,
+    ) -> Result<Vec<&ColumnLeaf>> {
+        let column_leaves = match self {
+            Projection::Columns(indices) => indices
+                .iter()
+                .map(|idx| &column_leaves.column_leaves[*idx])
+                .collect(),
+            Projection::InnerColumns(path_indices) => {
+                let paths: Vec<&Vec<usize>> = path_indices.values().collect();
+                paths
+                    .iter()
+                    .map(|path| {
+                        ColumnLeaves::traverse_path(&column_leaves.column_leaves, path).unwrap()
```

Comment:
This function is migrated from the old one. So I think it's safe. 

And the performance of `try_collect` is lower than `collect` because `try_collect` will not prealloc memory.